tee-worker/litentry/core/data-providers/src/moralis.rs

```
@@ -0,0 +1,268 @@
+// Copyright 2020-2024 Trust Computing GmbH.
+// This file is part of Litentry.
+//
+// Litentry is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use crate::{build_client, DataProviderConfig, Error, HttpError};
+use http::header::CONNECTION;
+use http_req::response::Headers;
+use itc_rest_client::{
+	http_client::{DefaultSend, HttpClient},
+	rest_client::RestClient,
+	RestGet, RestPath,
+};
+use litentry_primitives::Web3Network;
+use log::debug;
+use serde::{Deserialize, Serialize};
+use std::{
+	format, str,
+	string::{String, ToString},
+	thread, time, vec,
+	vec::Vec,
+};
+
+#[derive(Debug)]
+pub struct MoralisRequest {
+	path: String,
+	query: Option<Vec<(String, String)>>,
+}
+
+pub struct MoralisClient {
+	api_retry_delay: u64,
+	api_retry_times: u16,
+	client: RestClient<HttpClient<DefaultSend>>,
+}
+
+impl MoralisClient {
+	pub fn new(data_provider_config: &DataProviderConfig) -> Self {
+		let api_key = data_provider_config.moralis_api_key.clone();
+		let api_retry_delay = data_provider_config.moralis_api_retry_delay;
+		let api_retry_times = data_provider_config.moralis_api_retry_times;
+		let api_url = data_provider_config.moralis_api_url.clone();
```

It's out of this PR's scope, but sooner or later we'll need a refactoring to have a trait-based impl. The background is I see a lot of duplicate code, we could have something like `build_client_with auth` or `build_client_with retry` - or as fn parameters. 🤔

---

added fail fast boolean option in service layer, and trait in data provider layer now